File: proofload-java/src/test/java/io/github/matthewjones372/proofload/java/RunningTest.java
package io.github.matthewjones372.proofload.java;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.sun.net.httpserver.HttpServer;
import io.github.matthewjones372.proofload.Clock;
import io.github.matthewjones372.proofload.Offered;
import io.github.matthewjones372.proofload.RunResult;
import io.github.matthewjones372.proofload.Scenario;
import io.github.matthewjones372.proofload.SessionKey;
import io.github.matthewjones372.proofload.StepName;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.time.Duration;
import org.junit.jupiter.api.Test;

/**
 * A run rather than a construction: the claim is that a Java caller reaches the
 * whole path, and only a real target proves the accessors read what ran.
 */
class RunningTest {

    private static final StepName BROWSE = Steps.named("browse");

    private static final SessionKey<String> BODY = SessionKeys.of(String.class, "body");

    @Test
    void aJavaCallerRunsAScenarioAndReadsItsPercentiles() throws IOException {
        HttpServer server = HttpServer.create(new InetSocketAddress(0), 0);
        server.createContext("/products", exchange -> {
            exchange.getResponseHeaders().add("location", "/orders/1");
            exchange.sendResponseHeaders(200, 0);
            exchange.close();
        });
        server.start();

        try {
            var api = Https.baseUrl("http://localhost:" + server.getAddress().getPort());
            Scenario browsing = Scenarios.named("browsing")
                .exec(BROWSE, Https.capturing(api.get("/products").expecting(200), BODY, response -> response.header("location")))
                .build();

            RunResult result = Proofload.create()
                .run(Simulations.at(browsing, Rates.perSecond(20), Duration.ofMillis(500)));

            assertTrue(Results.ran(result, BROWSE));
            assertTrue(Results.count(result, BROWSE) > 0);
            assertEquals(Results.count(result, BROWSE), Results.ok(result, BROWSE));
            assertEquals(0, Results.failed(result, BROWSE));
            assertFalse(Results.p99(result, BROWSE).isNegative());
            assertTrue(Results.max(result, BROWSE).compareTo(Results.p50(result, BROWSE)) >= 0);
            assertFalse(Results.p95(result, BROWSE, Clock.ServiceTime).isNegative());

            Offered offered = Offereds.of(result);
            assertNotNull(offered);
            assertEquals(20.0, Offereds.asked(offered).getPerSecond());
            assertTrue(Offereds.left(offered).getPerSecond() > 0);
            assertFalse(Offereds.over(offered).isNegative());
            assertEquals(
                Offereds.left(offered).getPerSecond() / Offereds.asked(offered).getPerSecond(),
                offered.getShare()
            );
        } finally {
            server.stop(0);
        }
    }
}

File: proofload-scala/src/test/scala/io/github/matthewjones372/proofload/scala/RunningTest.scala
package io.github.matthewjones372.proofload.scala

import com.sun.net.httpserver.HttpServer
import io.github.matthewjones372.proofload.java.Results
import java.net.InetSocketAddress
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import _root_.scala.concurrent.duration.DurationInt

/**
 * A run rather than a construction: the claim is that a Scala caller reaches
 * the whole path, and only a real target proves the readers read what ran.
 */
class RunningTest:

  private val browse = step("browse")

  @Test
  def `a Scala caller runs a scenario and reads its percentiles`(): Unit =
    val server = HttpServer.create(InetSocketAddress(0), 0)
    server.createContext(
      "/products",
      exchange =>
        exchange.sendResponseHeaders(200, 0)
        exchange.close(),
    )
    server.start()

    try
      val api = http.baseUrl(s"http://localhost:${server.getAddress.getPort}")
      val browsing = scenario("browsing")(exec(browse, api.get("/products").expecting(200)))

      val result = Proofload().run(browsing.at(20.perSecond, over = 500.millis))

      assertTrue(result(browse).count > 0, "the run recorded nothing")
      assertEquals(result(browse).count, result(browse).ok)
      assertEquals(0L, result(browse).failed)
      assertTrue(result(browse).responseTime.p99 >= result(browse).responseTime.p50)
      assertTrue(result(browse).responseTime.max >= result(browse).responseTime.p99)
      assertTrue(result(browse).serviceTime.p95 >= _root_.scala.concurrent.duration.Duration.Zero)
      assertEquals(Results.p99(result, browse), asJava(result(browse).responseTime.p99))

      val offered = result.offered.get
      assertEquals(20.0, offered.asked.getPerSecond)
      assertTrue(offered.left.getPerSecond > 0.0, "nothing left, so nothing was offered")
      assertTrue(offered.over > _root_.scala.concurrent.duration.Duration.Zero)
      assertEquals(offered.left.getPerSecond / offered.asked.getPerSecond, offered.share)
    finally server.stop(0)
